### PR TITLE
feat: Import metadata from external instance

### PR DESCRIPTION
**Pull request by Karol-Stelmaczonek**
Ticket: https://oat-sa.atlassian.net/browse/AUT-3590

## What's Changed

- Add metametadata check & validation on import

## Dependencies PRs

- https://github.com/oat-sa/extension-tao-testqti/pull/2452
- https://github.com/oat-sa/extension-tao-testqti/pull/2454

## How to test
- Create import with metadata
- Run import
- If import class don't have proper metadata the import will fail

### CODE REVIEW

File: model/qti/ImportService.php [hunk 1]
```diff
@@ -107,6 +115,11 @@ public static function singleton()
      */
     protected $metadataImporter;
 
+    /**
+     * @var MetadataImporter Service to manage metaMetadata during package import
+     */
+    protected MetadataImporter $metaMetadataImporter;
+
```
Comment (bartlomiejmarszal) [inline model/qti/ImportService.php:122]: Why do you need this?

File: model/qti/ImportService.php [hunk 2]
```diff
@@ -176,6 +189,7 @@ protected function createRdfItem(core_kernel_classes_Class $itemClass, Item $qti
             throw new \common_Exception('Unable to save item');
         }
 
+
```
Comment (bartlomiejmarszal) [inline model/qti/ImportService.php:192]: More! More empty lines!

File: model/qti/metaMetadata/Exception.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Copyright (c) 2024 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+
+namespace oat\taoQtiItem\model\qti\metaMetadata;
+
+class Exception extends \common_Exception
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Exception.php:26]: Can we rename it?

Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Exception.php:26]: Move exception to use block
```suggestion
class Exception extends common_Exception
```

File: model/qti/metaMetadata/Exception.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2024 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Exception.php:23]: More! More empty lines!
```suggestion

```

File: model/qti/metaMetadata/Exporter.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+declare(strict_types=1);
+
+
+namespace oat\taoQtiItem\model\qti\metaMetadata;
+
+use oat\oatbox\service\ConfigurableService;
+
+class Exporter extends ConfigurableService
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Exporter.php:28]: [And if you need to build a second floor. You have a door already](https://www.youtube.com/watch?v=Uouj0eAxRc0)!

File: model/qti/metaMetadata/Importer.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            );
+        }
+
+        $xpath = new DOMXPath($manifest);
+        $xpath->registerNamespace('imsmd', self::NAMESPACE_LOM);
+        $xpath->registerNamespace('default', self::NAMESPACE_DEFAULT);
+
+        $metaMetadata = [];
+
+        $properties  = $xpath->query('//imsmd:metaMetadata/default:extension/default:customProperties/default:property');
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Importer.php:49]: Maybe we can move it to const?

File: model/qti/metaMetadata/Importer.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\qti\metaMetadata;
+
+use DOMDocument;
+use DOMXPath;
+use InvalidArgumentException;
+use oat\oatbox\service\ConfigurableService;
+
+class Importer extends ConfigurableService
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/Importer.php:30]: Can we change this name?

File: model/qti/metaMetadata/MetaMetadataService.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+declare(strict_types=1);
+
+
+namespace oat\taoQtiItem\model\qti\metaMetadata;
+
+use oat\oatbox\service\ConfigurableService;
+
+class MetaMetadataService extends ConfigurableService
+{
+    public const SERVICE_ID = 'taoQtiItem/metaMetadataService';
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/MetaMetadataService.php:30]: You can use `::class` when registering services
```suggestion
```

File: model/qti/metaMetadata/MetaMetadataService.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+declare(strict_types=1);
+
+
+namespace oat\taoQtiItem\model\qti\metaMetadata;
+
+use oat\oatbox\service\ConfigurableService;
+
+class MetaMetadataService extends ConfigurableService
```
Comment (bartlomiejmarszal) [inline model/qti/metaMetadata/MetaMetadataService.php:28]: Why do you need this class?

File: model/qti/ImportService.php [hunk 10]
```diff
@@ -903,4 +936,52 @@ private function getItemEventDispatcher(): UpdatedItemEventDispatcher
...
+                }
+            }
+
+            if (strlen($foundMetadata['checksum']) > 0
+                && $this->getChecksumGeneratorService()->getRangeChecksum($prop) !== $foundMetadata['checksum']
+            ) {
+                throw new MetaMetadataException(sprintf('Checksum mismatch for class property "%s"', $label));
+            }
+        }
+    }
```
Comment (bartlomiejmarszal) [inline model/qti/ImportService.php:986]: Extract this to new service

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
We should have a better control over metadata validation. I expect in future to not fail import when metadata fields are not matching. We could also import resources without metadata or import resources in class with metadata.

File: model/qti/ImportService.php [hunk 11]
```diff
@@ -903,4 +936,52 @@ private function getItemEventDispatcher(): UpdatedItemEventDispatcher
...
+        return new ChecksumGenerator($ontology);
+    }
+
+    public function validateClassMetadata($itemClass, array $metaMedataValues)
+    {
+        if (empty($metaMedataValues)) {
+            return;
+        }
+
+        $props = $itemClass->getProperties();
```
Comment (bartlomiejmarszal) [inline model/qti/ImportService.php:952]: Isn't this class a new class always so it will never has property values?

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
We should validate all metadata nested from all parents

File: model/qti/ImportService.php [hunk 12]
```diff
@@ -445,7 +461,8 @@ public function importQtiItem(
         $enableMetadataValidators = true,
         $itemMustExist = false,
         $itemMustBeOverwritten = false,
-        &$overwrittenItems = []
+        &$overwrittenItems = [],
+        $metaMedataValues = []
```
Comment (bartlomiejmarszal) [inline model/qti/ImportService.php:465]: You are taking all metadata from all package into `importQtiItem`. 
`importQtiItem` is importing particular item, no?

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
Can we have more control over metadata objects that we want to inject into resource?

File: model/qti/metadata/imsManifest/MetaMetadataValidator.php [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        foreach ($props as $prop) {
+            $label = $prop->getLabel();
+            $foundMetadata = array_filter($metaMedataValues, function ($metadataItem) use ($label) {
+                return $metadataItem['label'] === $label;
+            });
+            if (empty($foundMetadata)) {
+                throw new MetaMetadataException(sprintf('No metadata found for class property "%s"', $label));
+            }
+            if (count($foundMetadata) > 1) {
```
Comment (bartlomiejmarszal) [inline model/qti/metadata/imsManifest/MetaMetadataValidator.php:58]: If we don't find metadata corresponding to class properties we should not throw error.

**Review by tikhanovichA — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: model/qti/metadata/imsManifest/MetaMetadataExtractor.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+use DOMDocument;
+use DOMXPath;
+use InvalidArgumentException;
+
+class MetaMetadataExtractor
+{
+    private const NAMESPACE_LOM = 'http://ltsc.ieee.org/xsd/LOM';
+    private const NAMESPACE_DEFAULT = 'http://www.imsglobal.org/xsd/imscp_v1p1';
+    // phpcs:disable Generic.Files.LineLength
```
Comment (pnal) [inline model/qti/metadata/imsManifest/MetaMetadataExtractor.php:33]: you can probably break it in two lines instead, I guess it's valid

File: model/qti/metadata/imsManifest/MetaMetadataExtractor.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    private const META_METADATA_PROPERTIES_QUERY_STRING = '//imsmd:metaMetadata/default:extension/default:customProperties/default:property';
```
Comment (pnal) [inline model/qti/metadata/imsManifest/MetaMetadataExtractor.php:34]: AFAIK breaking it in two lines would also be legit
```suggestion
    private const META_METADATA_PROPERTIES_QUERY_STRING =
    '//imsmd:metaMetadata/default:extension/default:customProperties/default:property';
```

**Review by pnal — APPROVED** [review]
Good job!
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Pull request merged**